experimental/serde_encoding/src/lib.rs

```diff
@@ -0,0 +1,125 @@
+//
+// Copyright 2020 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+//! `serde` based serialization formats for inter-node communication.
+//!
+//! If your type is serializable using `serde` and implements `HandleVisit`, these formats can be
+//! used to send those types to other Oak nodes, including any handles they might contain.
+//!
+//! ```
+//! use oak::{
+//!     handle::HandleVisit,
+//!     io::{Decodable, Encodable, Receiver},
+//! };
+//! use serde::{Deserialize, Serialize};
+//! use serde_encoding::Bincoded;
+//!
+//! #[derive(Clone, Serialize, Deserialize, HandleVisit, Debug, PartialEq)]
+//! struct Test {
+//!     a: String,
+//!     b: Receiver<()>,
+//! }
+//!
+//! let test = Test {
+//!     a: "Hello, world!".to_string(),
+//!     b: Receiver::new(oak::ReadHandle { handle: 42 }),
+//! };
+//! let bincoded = Bincoded(test);
+//! let msg = bincoded.encode().unwrap();
+//!
+//! // Send `msg` to another node
+//!
+//! // On the recipient:
+//! let decoded: Bincoded<Test> = Bincoded::decode(&msg).unwrap();
+//!
+//! // Field values and handles are preserved
+//! assert_eq!(decoded.0, bincoded.0);
+//! ```
+
+use oak::{
+    handle::HandleVisit,
+    io::{Decodable, Encodable, Message},
+    OakError,
+};
+use serde::{Deserialize, Serialize};
+
+/// [`bincode`](https://crates.io/crates/bincode) format.
+pub struct Bincoded<T>(pub T);
+
+impl<T: Clone + Serialize + HandleVisit> Encodable for Bincoded<T> {
+    fn encode(&self) -> Result<Message, OakError> {
+        Ok(Message {
+            bytes: bincode::serialize(&self.0).map_err(|_| OakError::ProtobufEncodeError(None))?,
+            handles: oak::handle::extract_handles(&mut self.0.clone()),
+        })
+    }
+}
```

Comment:
Would it not be possible to have a generic fn to which we can pass a specific serialization fn (e.g. `bincode::serialize` or `serde_json::to_string`) and it runs that in parallel to the handle extraction? As long as that fn returns something that can be treated as a `&[u8]` it should be fine, right? Then we would not need wrapper types for each serialization mechanism.

Reply:
That's not a bad idea! Since we're keeping this mostly internal and we don't need to mix serializer that'll probably work fine, thanks!

Reply:
On second thought, we might actually need to keep the wrapper types around. The reason is that types like `InitWrapper` expect their inner types to be `Encodable + Decodable`, so a type like `InitWrapper<Bincoded<MySerdeStruct>, _>` is valid, but `InitWrapper<MySerdeStruct, _>` is not. This also holds the other way around, you can't have `MySerdeStruct<GrpcInvocation>` and have it be `Serializable + Deserialize` *or* `Encodable + Decodable`, but something like `MySerdeStruct<ProtoSerde<GrpcInvocation>>` would work (`ProtoSerde` would encode the proto as a `Vec<u8>`).

We could try and mitigate this by automatically deriving `Serialize + Deserialize` for all generated protos, but I'm not sure I'm a fan of that.

I'm starting to think the combination of serde and protocol buffers might introduce a lot of sharp edges, what do you think?